crates/matrix-sdk/src/send_queue.rs

```diff
@@ -950,13 +952,45 @@ impl SendHandle {
     }
 }
 
+/// From a given source of [`DependentQueuedEvent`], return only the most
+/// meaningful, i.e. the ones that wouldn't be overridden after applying the
+/// others.
+fn canonicalize_dependent_events(dependent: &[DependentQueuedEvent]) -> Vec<DependentQueuedEvent> {
+    let mut latest_edit = None;
```

Comment:
Hmm, this means making multiple iterations over the same list: one for redaction, then one for edits (and at most one more for reactions). I can get rid of the cloning for sure while still using a single plain loop, would that be ok?

Reply:
Did I misunderstand what the code does? Where do reactions come into play? Where are multiple iterations over the list? Are you talking about the future additions where we have more kinds?

Since both branches do the same thing you could it simplify to this as well:

```rust
let mut latest_edit = None;

for d in dependent {
    latest_edit = Some(d);
    match &d.kind {
        DependentQueuedEventKind::Edit { .. } => (),
        DependentQueuedEventKind::Redact => {
            // Shortcut and return the redaction; any other action would be meaningless,
            // since it'll end up with a redact.
            break;
        }
    }
}

latest_edit.map(|e| vec![e.clone()]).unwrap_or_default()
```

But now it's really obvious that you're searching for a redaction, and if you don't find one you take the last element, which the `find()` expresses neatly.

Reply:
> Did I misunderstand what the code does? Where do reactions come into play? Where are multiple iterations over the list? Are you talking about the future additions where we have more kinds?

Reactions come in another PR. Consider a future where we have reactions too: the `find()` isn't sufficient, because we really want *all* the reactions (so accumulating them), not only picking one reaction, and dramatically not picking a reaction over an edit or a redaction.

(Also it's not a `latest_edit` if it's a redaction :raised_eyebrow: )

Reply:
Have you looked at the new version where I avoid the cloning, until I get to add reactions there? It's good enough™ IMO, and a plain loop is quite expressive and simple.

Reply:
> (Also it's not a latest_edit if it's a redaction 🤨 )

:roll_eyes: 

>Have you looked at the new version where I avoid the cloning, until I get to add reactions there? It's good enough™ IMO, and a plain loop is quite expressive and simple.

Fine, I'll take another look at this once reactions come into play.